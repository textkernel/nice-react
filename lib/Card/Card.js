import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import props from '../props';

const Sub = ({ children, className }) => (<div className={ classNames('nice-card-sub', className)}>
    { children }
</div>);

class Card extends Component {
    static propTypes = {
        children: PropTypes.oneOfType([
            PropTypes.node,
            PropTypes.array
        ]),
        className: PropTypes.string,
        context: PropTypes.oneOf(props.cardContexts),
        controls: PropTypes.arrayOf(PropTypes.shape({
            disabled: PropTypes.bool,
            href: PropTypes.string,
            target: PropTypes.string,
            onClick: PropTypes.func,
            label: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.node
            ]).isRequired,
            className: PropTypes.string
        })),
        style: PropTypes.object
    }

    static defaultProps = {
        children: null,
        className: null,
        context: null,
        controls: [],
        style: null
    }

    renderTitle() {
        const { title } = this.props;

        if (!title) {
            return null;
        }

        return (<div className="nice-card-title">
            <h3>{ title }</h3>
        </div>);
    }

    renderImage() {
        const { image } = this.props;

        if (!image) {
            return null;
        }

        return (<img
            className="nice-card-image"
            src={ image }
        />);
    }

    renderControls() {
        const { controls } = this.props;

        if (!controls || !Array.isArray(controls) || !controls.length) {
            return null;
        }

        return (<div className="nice-card-controls">
            { controls.map((control, i) => {
                const { disabled, href = '#', label, target, onClick } = control;

                return (<a
                    key={ i }
                    href={ href }
                    target={ target }
                    onClick={ onClick }
                    className={ classNames('nice-card-control', {
                        disabled
                    }) }
                >
                    { label }
                </a>);
            }) }
        </div>);
    }

    render() {
        const { children, className, context, controls, style } = this.props;

        return (
            <div className={ classNames('nice-card', className, {
                [context]: context,
                'has-controls': Array.isArray(controls) && controls.length
            }) } style={ style }>
                { this.renderTitle() }
                { this.renderImage() }
                { React.Children.map(children, (child, i) => {
                    if (!child.type || typeof child.type === 'string') {
                        return (<div className="nice-card-content" key={ i }>
                            { child }
                        </div>);
                    }
                    return child;
                }) }
                { this.renderControls() }
            </div>
        );
    }
}

Card.Sub = Sub;

export default Card;
