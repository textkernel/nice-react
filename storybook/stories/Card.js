import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, number, select } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';
import props from '../../src/props';

import { Card } from 'nice-react';

const stories = storiesOf('Cards', module);

stories.addDecorator((story, context) => withInfo('')(story)(context));
stories.addDecorator(withKnobs);

const onClick = e => {
    e.preventDefault();
};

stories.add('Default', () => {
    const title = text('Title', 'This is a card');
    const content = text('Content', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur ullamcorper nulla ac quam hendrerit pharetra.');
    const context = select('Context', props.cardContexts);
    const subs = number('Sub items', 2);
    const controls = boolean('Has controls', true);
    const image = boolean('Image', true) ? '//placeholdit.imgix.net/~text?txtsize=33&txt=300%C3%97150&w=300&h=150' : null;

    return (<Card
        title={ title }
        context={ context }
        controls={ controls ? [{
            label: 'OK',
            className: 'icon fa-check',
            onClick
        }, {
            label: 'Dismiss',
            disabled: true,
            onClick
        }] : null }
        image={ image }
        style={{
            float: 'none',
            width: '300px'
        }}
    >
        <Card.Content>
            { content }
        </Card.Content>
        {
            Array(subs).fill(0).map((v, k) => (
                <Card.Sub className="icon fa-reply" key={ k }>
                    In semper vel lectus sed finibus. Donec in erat risus. Aenean in lectus massa. Phasellus id malesuada arcu, vel ultricies enim.
                </Card.Sub>
            ))
        }
    </Card>);
});
