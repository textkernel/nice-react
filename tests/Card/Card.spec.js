import React from 'react';
import Card from '../../lib/Card';

const onClick = () => {};

describe('<Card />', () => {
    it('should render default Card without subs', () => {
        const wrapper = shallow(<Card
            title="This is a card"
            controls={ [{
                label: 'OK',
                className: 'icon fa-check',
                onClick
            }, {
                label: 'Dismiss',
                disabled: true,
                onClick
            }] }
            image="//placeholdit.imgix.net/~text?txtsize=33&txt=300%C3%97150&w=300&h=150"
            style={{
                float: 'none',
                width: '300px'
            }}
        >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur
            ullamcorper nulla ac quam hendrerit pharetra.
        </Card>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render Card with subs, without controls', () => {
        const wrapper = shallow(<Card
            title="This is a card with subs"
            context="brand"
            style={{
                float: 'none',
                width: '300px'
            }}
        >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur
            ullamcorper nulla ac quam hendrerit pharetra.
            <Card.Sub className="icon fa-reply">
                In semper vel lectus sed finibus. Donec in erat risus. 
                Aenean in lectus massa. Phasellus id malesuada arcu, vel ultricies enim.
            </Card.Sub>
            <Card.Sub className="icon fa-reply">
                    In semper vel lectus sed finibus. Donec in erat risus. 
                    Aenean in lectus massa. Phasellus id malesuada arcu, vel ultricies enim.
            </Card.Sub>
        </Card>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render Card with custom classnames/styles', () => {
        const wrapper = shallow(
            <Card
                title="This is a card with subs"
                context="brand"
                className="my-classname"
                style={{
                    float: 'none',
                    width: '300px'
                }}
                controls={ [{
                    label: 'OK',
                    className: 'icon fa-check',
                    style: { color: 'green' },
                    onClick
                }]}
            >
                <Card.Content className="content-classname" style={{ color: 'red' }}>
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur
                    ullamcorper nulla ac quam hendrerit pharetra.
                </Card.Content>

                <Card.Sub className="icon fa-reply" style={{ backgroundColor: 'blue' }}>
                    In semper vel lectus sed finibus. Donec in erat risus.
                    Aenean in lectus massa. Phasellus id malesuada arcu, vel ultricies enim.
                </Card.Sub>
            </Card>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
