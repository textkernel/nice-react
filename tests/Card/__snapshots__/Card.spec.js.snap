// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render Card with subs, without controls 1`] = `
<div
  className="nice-card brand"
  style={
    Object {
      "float": "none",
      "width": "300px",
    }
  }
>
  <div
    className="nice-card-title"
  >
    <h3>
      This is a card with subs
    </h3>
  </div>
  <div
    className="nice-card-content"
    key="0/.0"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur ullamcorper nulla ac quam hendrerit pharetra.
  </div>
  <Sub
    className="icon fa-reply"
    key=".1"
  >
    In semper vel lectus sed finibus. Donec in erat risus. Aenean in lectus massa. Phasellus id malesuada arcu, vel ultricies enim.
  </Sub>
  <Sub
    className="icon fa-reply"
    key=".2"
  >
    In semper vel lectus sed finibus. Donec in erat risus. Aenean in lectus massa. Phasellus id malesuada arcu, vel ultricies enim.
  </Sub>
</div>
`;

exports[`<Card /> should render default Card without subs 1`] = `
<div
  className="nice-card has-controls"
  style={
    Object {
      "float": "none",
      "width": "300px",
    }
  }
>
  <div
    className="nice-card-title"
  >
    <h3>
      This is a card
    </h3>
  </div>
  <img
    className="nice-card-image"
    src="//placeholdit.imgix.net/~text?txtsize=33&txt=300%C3%97150&w=300&h=150"
  />
  <div
    className="nice-card-content"
    key="0/.0"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur ullamcorper nulla ac quam hendrerit pharetra.
  </div>
  <div
    className="nice-card-controls"
  >
    <a
      className="nice-card-control"
      href="#"
      key="0"
      onClick={[Function]}
    >
      OK
    </a>
    <a
      className="nice-card-control disabled"
      href="#"
      key="1"
      onClick={[Function]}
    >
      Dismiss
    </a>
  </div>
</div>
`;
